keyboards/keychron/common/matrix.c

```diff
@@ -61,33 +61,51 @@ static inline uint8_t readMatrixPin(pin_t pin) {
     }
 }
 
-static void shiftOut(uint8_t dataOut) {
-    for (uint8_t i = 0; i < 8; i++) {
-        if (dataOut & 0x1) {
-            setPinOutput_writeHigh(DATA_PIN);
-        } else {
-            setPinOutput_writeLow(DATA_PIN);
+// At 3.6V input, three nops (37.5ns) should be enough for all signals
+#define small_delay() __asm__ __volatile__("nop;nop;nop;\n\t" ::: "memory")
```

Comment:
Since we don't actually know the voltage that the chip runs at, we should make this a bit more. Three `nop`s work on my Keychron Q3, but there's no guarantee that it will work on everyone's Keychron Q*. The absolute maximum according to the data sheet is 125ns, so 10 `nop`s at 80 MHz. I don't know if we want to go with the safest option or just go with 4 `nop`s as an "it'll probably work", I'll leave that up to you.

Reply:
Of course if you could find out the actual voltage the chip runs at that would be ideal.

Reply:
Hi, the chips of all the Keychron Q* and V* keyboards run at 3.3V.

Reply:
I tested it on q1/q3/q5/q6/q10 and v1/v3/v5/ and it worked normally.

Reply:
Given that it can vary slightly depending on connected host, I can assume that its 100% not enough testing for it to say its "working" for the general population.

Reply:
4.35V is minimum allowable voltage for USB2.0 at the UFP-side. Suggest confirmation that things are fine -- on-board 3.3V LDOs and the like should ensure that this is still handled, but would be good to confirm.

Reply:
At 3.3V this should definitely be at least 4 `nop`s. Technically the only correct thing would be to use the timings for 2V since there are none for 3.3V specifically, but we can just interpolate and it will probably work.

Reply:
@tzarc Hi, we can learn from the datasheet of this chip (STM32L432KBU6) that the USB supply voltage of this chip can be between 3.0V and 3.6V.

In addition, the onboard 3.3V LDO we used is the ME6202A series of MicrOne. It has excellent input stability and the maximum input voltage up to 6V, and the operating temperature is from -25℃ to+85℃, which is enough to support MCU operation.

The IO supply voltage of CKLED2001 driving all LEDs is directly from VBUS, which has little effect on the 3.3V voltage.

I have put all the relevant IC datasheets in OneDrive. Welcome to download and check.

Reply:
@lalalademaxiya1 I tried flashing this branch to my keychron q10, and while it worked just fine, the keyboard disappeared from VIA on the device authorization dialog. Flashing back the latest firmware from keychron website brought it back (https://cdn.shopify.com/s/files/1/0059/0630/1017/files/q10_us_knob_v1.2.bin?v=1671504855) .

I wonder if this is expected? It might be purely user error on my side; I flashed using `make keychron/q10/ansi_encoder:flash` after checking out this branch

Reply:
Make sure that you use the via or keychron keymaps instead of the default one @ggayan ;)